packages/upload-api/src/ucan/conclude.js

```diff
@@ -0,0 +1,163 @@
+          /** @type {API.UnknownLink} */
+          // @ts-expect-error ts does not know how to get this
+          const blobAllocateTaskCid = cap.nb.url['ucan/await'][1]
+          const blobAllocateTaskGet = await tasksStorage.get(blobAllocateTaskCid)
+          if (blobAllocateTaskGet.error) {
+            return blobAllocateTaskGet
+          }
+
+          /** @type {API.BlobAllocate} */
+          // @ts-expect-error ts does not know how to get this
+          const allocateCapability = blobAllocateTaskGet.ok.capabilities.find(
+            // @ts-expect-error ts does not know how to get this
+            (/** @type {API.BlobAllocate} */ allocateCap) =>
+              equals(allocateCap.nb.blob.digest, cap.nb.body.digest) && allocateCap.can === W3sBlob.allocate.can
+          )
+          
+          const blobAccept = await W3sBlob.accept
+            .invoke({
+              issuer: id,
+              audience: id,
+              with: id.toDIDKey(),
+              nb: {
+                blob: cap.nb.body,
+                exp: Number.MAX_SAFE_INTEGER,
+                space: allocateCapability.nb.space,
+                _put: {
+                  'ucan/await': ['.out.ok', ranInvocation.link()],
+                },
+              },
+              expiration: Infinity,
+            })
+            .delegate()
+
+          return tasksScheduler.schedule(blobAccept)
+        })
+    )
+
+    const scheduleErrors = scheduleRes.filter((res) => res.error)
+    if (scheduleErrors.length && scheduleErrors[0].error) {
+      return {
+        error: scheduleErrors[0].error,
+      }
+    }
+
+    return {
+      ok: { time: Date.now() },
+    }
+  })
+
+/**
+ * @param {import('multiformats').UnknownLink} cid
+ * @param {IterableIterator<import('@ucanto/interface').Block<unknown, number, number, 1>>} blocks
+ * @returns {Promise<import('@ucanto/interface').Result<Uint8Array, DecodeBlockOperationFailed>>}
+ */
+export const findBlock = async (cid, blocks) => {
+  let bytes
+  for (const b of blocks) {
+    if (b.cid.equals(cid)) {
+      bytes = b.bytes
+    }
+  }
+  if (!bytes) {
+    return {
+      error: new DecodeBlockOperationFailed(`missing block: ${cid}`),
+    }
+  }
+  return {
+    ok: bytes,
+  }
+}
+
+/**
+ * @param {import('@ucanto/interface').Invocation} concludeFx
+ */
+export function getConcludeReceipt(concludeFx) {
+  const receiptBlocks = new Map()
+  for (const block of concludeFx.iterateIPLDBlocks()) {
+    receiptBlocks.set(`${block.cid}`, block)
+  }
+  return Receipt.view({
+    // @ts-expect-error object of type unknown
+    root: concludeFx.capabilities[0].nb.receipt,
+    blocks: receiptBlocks,
+  })
+}
+
+/**
+ * @param {API.Signer} id
+ * @param {API.Verifier} serviceDid
+ * @param {API.Receipt} receipt
+ */
+export function createConcludeInvocation(id, serviceDid, receipt) {
+  const receiptBlocks = []
+  for (const block of receipt.iterateIPLDBlocks()) {
```

Comment:
Might be a good idea to stick cids into second array to avoid second iteration in `facts`.